services/worker-manager/src/providers/azure.js

```diff
@@ -387,19 +383,229 @@ class AzureProvider extends Provider {
+   */
+  async handleOperation({op, errors}) {
+    let req, resp;
+    try {
+      // TODO: we could be smarter about getting operation status
+      // see here: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations
+      req = new msRestJS.WebResource(op, 'GET');
+      // sendLongRunningRequest polls until finished but this is just reading
+      // the status of an operation so it shouldn't block long
+      // it's ok if we hit an error here, that will trigger resource teardown
+      resp = await this._enqueue('opRead', () => this.restClient.sendLongRunningRequest(req));
+    } catch (err) {
+      // Rest API has different error semantics than the SDK
+      if (_.has(resp, 'status') && resp.status === 404) {
+        // operation not found because it has either expired or does not exist
+        // nothing more to do
+        return false;
+      }
+      throw err;
+    }
+
+    let body = resp.parsedBody;
+    if (body) {
+      // status is guaranteed to exist if the operation was found
+      if (body.status === 'InProgress') {
+        return true;
+      }
+      if (body.error) {
+        errors.push({
+          kind: 'operation-error',
+          title: 'Operation Error',
+          description: body.error.message,
+          extra: {
+            code: body.error.code,
+          },
+          notify: this.notify,
+          WorkerPoolError: this.WorkerPoolError,
+        });
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * provisionResource generically provisions individual resources
+   * Handles cases where:
+   *  we have not yet created a resource and need to create one,
+   *  we have requested a resource but it is not ready,
+   *  we have a resource ready to go
+   *
+   * worker: the worker for which the resource is being provisioned
+   * client: the Azure SDK client for the resource
+   * resourceType: the short name used to identify the resource in providerData
+   * resourceConfig: configuration to be passed to the SDK for resource creation
+   * modifyFn: a function (worker, resource) that takes the worker and the created
+   *   resource, allowing the worker to be modified.
+   */
+  async provisionResource({worker, client, resourceType, resourceConfig, modifyFn}) {
+    if (!_.has(worker.providerData, resourceType)) {
+      throw new Error(`Error provisioning worker: providerData does not contain resourceType ${resourceType}`);
+    }
+    let typeData = worker.providerData[resourceType];
+
+    // we have no id, so we try to lookup resource by name
+    if (!typeData.id) {
+      try {
+        let resource = await this._enqueue('query', () => client.get(
+          worker.providerData.resourceGroupName,
+          typeData.name,
+        ));
+        // we found the resource
+        await worker.modify(w => {
+          w.providerData[resourceType].id = resource.id;
+          modifyFn(w, resource);
+        });
```

Comment:
do you walso want to update `typeData.id` here, or does that just get seen on the next go-round?

Reply:
My impression is that `typeData.id` refers to the same object (?) and so it is effectively getting updated here.

Reply:
Not necessarily -- the `modify` function might have to reload the worker object, and iirc ends up rewriting the properties anyway.